errors/errors.go

```diff
@@ -0,0 +1,300 @@
+		err.httpCode = HTTPStatusFromCode(grpcStatusCode)
+	}
+}
+
+// WithResourceInfo used to pass ResourceInfo to the Error struct.
+func WithResourceInfo(resourceInfo *ResourceInfo) Option {
+	return func(e *Error) {
+		e.resourceInfo = resourceInfo
+	}
+}
+
+// WithDescription used to pass a description
+// to the Error struct.
+func WithDescription(description string) Option {
+	return func(e *Error) {
+		e.description = description
+	}
+}
+
+// WithMetadata used to pass a Metadata[string]string
+// to the Error struct.
+func WithMetadata(md map[string]string) Option {
+	return func(e *Error) {
+		e.metadata = md
+	}
+}
+
+func newErrorInfo(reason string, md map[string]string) *errdetails.ErrorInfo {
+	return &errdetails.ErrorInfo{
+		Domain:   domain,
+		Reason:   reason,
+		Metadata: md,
+	}
+}
+
+func newResourceInfo(rid *ResourceInfo, err error) *errdetails.ResourceInfo {
+	owner := owner
+	if rid.Owner != "" {
+		owner = rid.Owner
+	}
+	return &errdetails.ResourceInfo{
+		ResourceType: rid.Type,
+		ResourceName: rid.Name,
+		Owner:        owner,
+		Description:  err.Error(),
+	}
+}
+
+// *** GRPC Methods ***
+
+// GRPCStatus returns the gRPC status.Status object.
+func (e *Error) GRPCStatus() *status.Status {
+	var stErr error
+	ste := status.New(e.grpcStatusCode, e.description)
+	if e.resourceInfo != nil {
+		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata), newResourceInfo(e.resourceInfo, e.err))
+	} else {
+		ste, stErr = ste.WithDetails(newErrorInfo(e.reason, e.metadata))
+	}
+	if stErr != nil {
+		return status.New(codes.Internal, fmt.Sprintf("failed to create gRPC status message: %v", stErr))
+	}
+
+	return ste
+}
+
+// *** HTTP Methods ***
+
+// ToHTTP transforms the supplied error into
+// a GRPC Status and then Marshals it to JSON.
+// It assumes if the supplied error is of type Error.
+// Otherwise, returns the original error.
+func (e *Error) ToHTTP() (int, []byte) {
+	resp, err := protojson.Marshal(e.GRPCStatus().Proto())
+	if err != nil {
+		errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
+		return http.StatusInternalServerError, errJSON
+	}
+
+	return e.httpCode, resp
+}
+
+// HTTPCode returns the value of the HTTPCode property.
+func (e *Error) HTTPCode() int {
+	if e == nil {
+		return http.StatusOK
+	}
+
+	return e.httpCode
+}
+
+// JSONErrorValue implements the errorResponseValue interface (used by `github.com/dapr/dapr/pkg/http`).
+func (e *Error) JSONErrorValue() []byte {
+	b, err := protojson.Marshal(e.GRPCStatus().Proto())
+	if err != nil {
+		errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
+		return errJSON
+	}
+	return b
+}
+
+// HTTPStatusFromCode converts a gRPC error code into the corresponding HTTP response status.
+// https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/errors.go#L15
+// See: https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+func HTTPStatusFromCode(code codes.Code) int {
```

Comment:
Please move this and the method below to a separate package (maybe `dapr/kit/grpccodes`? naming is hard - you pick)

We can then update dapr/dapr to use that package, but should be on its own